### PR TITLE
Fix/condo/sberdoma 721/resident api fixes soft delete

### PR DESCRIPTION
**Pull request by AntonAL**
Only soft-delete update-operation is allowed for resident.
Resident can now be updated from current user in session, which is equal to Resident's user.

### CODE REVIEW

File: apps/condo/domains/resident/access/Resident.js [hunk 1]
```diff
@@ -18,8 +19,24 @@ async function canReadResidents ({ authentication: { item: user } }) {
 async function canManageResidents ({ authentication: { item: user }, originalInput, operation, itemId, context }) {
     if (!user) return false
     if (user.isAdmin) return true
-    if (operation === 'create' && user.type === RESIDENT) {
-        return true
+    if (user.type === RESIDENT) {
+        if (operation === 'create') return true
```
Comment (pahaz) [inline apps/condo/domains/resident/access/Resident.js:23]: We have register resident ? We don’t need to create it

File: apps/condo/domains/resident/access/Resident.js [hunk 2]
```diff
@@ -18,8 +19,24 @@ async function canReadResidents ({ authentication: { item: user } }) {
...
+        // Only soft-delete is allowed for current resident
+        if (operation === 'update') {
```
Comment (pahaz) [inline apps/condo/domains/resident/access/Resident.js:25]: It’s better to check only deletedAt ticket

Comment (toplenboren) [reply apps/condo/domains/resident/access/Resident.js:25]: You also should check if the deletedAt is the only key in the query

Right now I can update some properties while deleting an object

File: apps/condo/domains/resident/schema/Resident.test.js [hunk 3]
```diff
@@ -412,5 +424,32 @@ describe('Resident', () => {
             expect(objUpdated.updatedAt).toMatch(DATETIME_RE)
             expect(objUpdated.updatedAt).not.toEqual(objUpdated.createdAt)
         })
+
+        it('can be soft-deleted using update operation by current user with type resident', async () => {
```
Comment (toplenboren) [inline apps/condo/domains/resident/schema/Resident.test.js:428]: Try also to delete user while updating it

File: apps/condo/domains/resident/access/Resident.js (hunk 1 above)
Comment (AntonAL) [reply apps/condo/domains/resident/access/Resident.js:23]: Hmm, I thought, that access control hooks are being called, when executing custom mutation (with passing not modified context). I see, that they are bypassed.

Comment (AntonAL) [reply apps/condo/domains/resident/access/Resident.js]: Removed in `dc1fc488`.

File: apps/condo/domains/resident/schema/Resident.test.js (hunk 3 above)
Comment (AntonAL) [reply apps/condo/domains/resident/schema/Resident.test.js:392]: Added extra checks in `737cbaa1`

File: apps/condo/domains/resident/access/Resident.js (hunk 2 above)
Comment (AntonAL) [reply apps/condo/domains/resident/access/Resident.js:25]: Added extra checks in `737cbaa1`

File: apps/condo/domains/resident/access/Resident.js (hunk 1 above)
Comment (AntonAL) [reply apps/condo/domains/resident/access/Resident.js]: Fixed in `058686b1`

File: apps/condo/domains/resident/access/Resident.js [hunk 4]
```diff
@@ -18,8 +20,28 @@ async function canReadResidents ({ authentication: { item: user } }) {
...
+            }
+            const [resident] = await Resident.getAll(context, { id: itemId })
+            if (!resident) {
+                return false
+            }
+            if (user.id !== resident.user.id) {
+                return false
+            }
+            if (
+                originalInput.deletedAt !== null &&
```
Comment (dkoviazin) [inline apps/condo/domains/resident/access/Resident.js:37]: Maybe there is a more simple way to check - without checking Object.keys(originalInput).length === 3
At least we can define some function to check if update action was in fact softDelete to use this later

**Review by dkoviazin — APPROVED** [review]

File: apps/condo/domains/resident/access/Resident.js (hunk 4 above)
Comment (AntonAL) [reply apps/condo/domains/resident/access/Resident.js]: Extracted in `9eb94e62`

File: packages/@core.keystone/access.js [hunk 5]
```diff
@@ -54,6 +55,15 @@ const readOnlyField = {
     update: false,
 }
 
+const isSoftDelete = (originalInput) => {
+    return (
+        originalInput.deletedAt !== null &&
```
Comment (pahaz) [inline packages/@core.keystone/access.js:60]: newId is also related to softDelete

**Review by pahaz — APPROVED** [review]

File: packages/@core.keystone/access.js (hunk 5 above)
Comment (AntonAL) [reply packages/@core.keystone/access.js]: Added in `066ce93b`

**Pull request merged**